### PR TITLE
Add Slycot exception and warning classes

### LINKED ISSUE #99: Replace ValueError with info attribute with SlycotError
See https://github.com/python-control/Slycot/pull/84#issuecomment-611991035

A common pattern in the Slycot code is

```
        e = ValueError('the number of eigenvalues to be assigned is less than the number of possibly assignable eigenvalues')
        e.info = info
        raise e
```

The idea is to have a SlycotError class, so that this can be:

```
    raise SlycotError(msg, info)
```
instead - that way one can't forget the .info attribute.

I'd have had `SlycotError` inherit from `RuntimeError`, but for backward compatibility it should probably be `ValueError`.

### PR DESCRIPTION
**Pull request by bnavigator**
As proposed in #99, I am introducing the new Exceptions `SlycotError`, `SlycotParameterError` and `SlycotArithmeticError`. So to never forget the `.info` attribute again!

Fixes #99

### CODE REVIEW

File: slycot/analysis.py [hunk 1]
```diff
@@ -121,28 +124,21 @@ def ab01nd(n,m,A,B,jobz='N',tol=0,ldwork=None):
     if jobz == 'N':
         out = _wrapper.ab01nd_n(n,m,A,B,tol=tol,ldwork=ldwork)
         if out[-1] < 0:
-            error_text = "The following argument had an illegal value: "+arg_list[-out[-1]-1]
-            raise ValueError(error_text)
+            raise SlycotParameterError(None, out[-1], arg_list)
```
Comment (repagh) [inline slycot/analysis.py:127]: Looks a bit ugly and hackish

Would it be an idea to use named parameters?
```
raise SlycotParameterError(info=out[-1], arglist=arg_list)
```
Or call 
```
SlycotParameterError(arg_list, out[-1])
```
then detect that the first parameter is not a string and compose a message?

File: slycot/analysis.py [hunk 2]
```diff
@@ -679,11 +660,11 @@ def ab09ad(dico,job,equil,n,m,p,A,B,C,nr=None,tol=0,ldwork=None):
     else:
         ordsel = 'F'
     if dico != 'C' and dico != 'D':
-        raise ValueError('Parameter dico had an illegal value')
+        raise SlycotParameterError('Parameter dico had an illegal value', -1)
```
Comment (repagh) [inline slycot/analysis.py:663]: Here you are essentially inventing new info numbers. What will they be used for?

Comment (repagh) [reply slycot/analysis.py:663]: Ah no, I see, they correspond to the SLICOT parameter number. OK, ignore my remark.

File: slycot/analysis.py (hunk 1 above)
Comment (bnavigator) [reply slycot/analysis.py:127]: Thanks for looking at this!

I wanted to make it consistent with the `("msg", info)` order of parameters. 

I personally don't like the overloading of a single constructor but that would be possible.

Another option: Have an additional subclass derived from `SlycotParameterError` with `SlycotGenericParameterError(arg_list, info)`. Maybe there is a better idea for the name, though.

Or get rid of the ever same
```python

if info < 0:
    raise SlycotParameterError(....)
```

and replace it with a call of

```python

def check_illegal_parameter(arg_list, info):
    if info < 0:
         message = ("The following argument had an illegal value: {}"
                    "".format(arg_list[-info-1]))
         raise SlycotParameterError(message, info)
```

Comment (repagh) [reply slycot/analysis.py:127]: That last solution is the way to go!

Comment (repagh) [reply slycot/analysis.py:127]: I am getting another idea, can we format the messages from the "Raises" section in the functions docstring? That would force good docstrings, and eliminate the typing!

```python
def verify_slycot_return(arg_list, docstring, info):
    if info < 0:
         message = ("The following argument had an illegal value: {}"
                    "".format(arg_list[-info-1]))
         raise SlycotParameterError(message, info)
    elif info > 0:
         # process the docstring for the error message
         # ....
```

Comment (repagh) [reply slycot/analysis.py:127]: p.s. I am working on this now...

**Comment by repagh** [comment]
Ow. That went quicker than I thought. It seems I was able to push on your branch, if hope you like the modification, if not, ignore/revert. 

One of the advantages is that this reduces the number of lines; fewer lines - better coverage.

**Comment by bnavigator** [comment]
I like it, but right now the raise from docstring always raises `SlycotParameterError` even when it should be `SlycotArithmeticError`. Working on this right now

**Comment by repagh** [comment]
I think I accidentally did a wrong force-push, trying to fix the loose ampersand. :-(

**Comment by bnavigator** [comment]
Yeah, you discarded my edits. No worries, will fix that in my next commit. Probably need to rebase a little bit.

**Comment by bnavigator** [comment]
> Look, nice improvement:
> 
> ```
> coverage/coveralls — Coverage increased (+25.1%) to 68.396%
> ```

It's cheating though because we now parse stuff that is not counted by coverage. ;)

**Comment by repagh** [comment]
Nice work, you lifted it to a whole new level.

Some additional points
- I see you removed the parsing check from test_sb.py. I worked on an easy routine to verify that a docstring can be parsed for the exceptions that we think are in there. That would offset the "cheating"? 
- I am getting these complaints from git that the recent commit messages have a & and a <, fix or ignore?

**Comment by bnavigator** [comment]
Yes the original parsing check did not work anymore as the new parser does not return the complete list.

Will fix the commit messages in the next rebase, where I also fix your "merge" commit. Right now I am getting gihub connection errors on git.

I am working on extending this to "Warns" also. There are quite a few routines that use the info parameter for warnings but want to return partial results.

File: slycot/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    SlycotArithmeticError:
+    Info is 2, which is between 1 and 4
+    >>> fun(4)
+    SlycotArithmeticError:
+    4 is in [4,    1.2e+02)!
+    >>> fun(120)
+    SlycotResultWarning:
+    120 is too large
+    """
+    if docstring:
```
Comment (repagh) [inline slycot/exceptions.py:212]: shouldn't we check info > 0?
Now we are parsing the docstring after each calculation.

File: slycot/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        exception, message = _parse_docsection("Raises", docstring, checkvars)
+        if exception and message:
+            fmessage = '\n' + message.format(**checkvars).strip()
+            raise globals()[exception](fmessage, info)
+
+        warning, message = _parse_docsection("Warns", docstring, checkvars)
+        if warning and message:
+            fmessage = '\n' + message.format(**checkvars).strip()
+            warn(globals()[warning](fmessage, info))
+
```
Comment (repagh) [inline slycot/exceptions.py:224]: and if we don't find a warning or message, and info > 0, we should return a catch-all exception, with just the info number. Better safe than sorry.

File: slycot/math.py [hunk 5]
```diff
@@ -584,12 +556,9 @@ def mb03wd(job, compz, n, ilo, ihi, iloz, ihiz, H, Q, ldwork=None):
...
-    if info < 0:
-        e = ValueError(
-                "Argument '{}' had an illegal value".format(arg_list[-info-1]))
-        e.info = info
-        raise e
-    elif info > 0:
+    raise_if_slycot_error(info, arg_list)
+
+    if info > 0:
         warnings.warn(("failed to compute all the eigenvalues {ilo} to {ihi} "
```
Comment (repagh) [inline slycot/math.py:562]: can also be replaced?

File: slycot/math.py [hunk 6]
```diff
@@ -609,63 +578,78 @@ def mb05md(a, delta, balanc='N'):
...
+        appear consecutively with the eigenvalue having positive
+        imaginary part first.
+
+    Warns
+    ------
+    SlycotResultWarning : e
+        :0 < e.info <=n:
+            the QR algorithm failed to compute all
+            the eigenvalues; no eigenvectors have been computed;
+            w[{e.info:n}] contains eigenvalues which have converged;
```
Comment (repagh) [inline slycot/math.py:646]: [{e.info}:{n}]

File: slycot/exceptions.py (hunk 3 above)
Comment (bnavigator) [reply slycot/exceptions.py:212]: Custom messages with info < 0 should be possible

File: slycot/math.py (hunk 5 above)
Comment (bnavigator) [reply slycot/math.py:562]: Yes, there are a lot of functions which can now be simplified. I am currently going through 
`analysis.py`.

I wanted to do the docstring cleanup (#100) in a separate PR later, but now with all the modifications in the docstrings already, which is not that easy to review it is a good thing to do much work on that now.

File: slycot/exceptions.py (hunk 4 above)
Comment (bnavigator) [reply slycot/exceptions.py:224]: I agree, that now with the possibility to mention warnings, a catchall can be implemented.

File: slycot/math.py [hunk 7]
```diff
@@ -18,6 +18,8 @@
 #       MA 02110-1301, USA.
 
 from . import _wrapper
+from .exceptions import raise_if_slycot_error, SlycotArithmeticError
```
Comment (repagh) [inline slycot/math.py:21]: SlycotArithmeticError no longer needed

File: slycot/math.py [hunk 8]
```diff
@@ -552,22 +520,22 @@ def mb03wd(job, compz, n, ilo, ihi, iloz, ihiz, H, Q, ldwork=None):
...
-    ValueError : e
-            e.info contains information about the exact type of exception
-
+    Warns
+    -----
+    SlycotResultWarning
+        :info > 0:
+            failed to compute all the eigenvalues {ilo} to {ihi}
+            in a total of 30*({ihi}-{ilo}+1) iterations
+            the elements Wr{{info}:{ihi}] contains those
```
Comment (repagh) [inline slycot/math.py:537]: Wr{{info}:{ihi}]  to Wr[{info}:{ihi}]

File: slycot/exceptions.py (hunk 3 above)
Comment (repagh) [reply slycot/exceptions.py:224]: ok, but I am worried about performance
can we do
```python
    if docstring and (iwarn or info):
        ....
```
then we don't parse the docstring upon succesful calculation. Might matter to someone who is calculating a lot of LQR controllers or the like.

File: slycot/synthesis.py [hunk 9]
```diff
@@ -286,20 +279,25 @@ def sb02md(n,A,G,Q,dico,hinv='D',uplo='U',scal='N',sort='S',ldwork=None):
...
+        :info = 4:
+            The Hamiltonian or symplectic matrix H has less
+            than n stable eigenvalues;
+        :info = 5:
+            The n-th order system of linear algebraic
+            equations, from which the solution matrix X would
+            be obtained, is singular to working precision.
 
 
         Example
```
Comment (repagh) [inline slycot/synthesis.py:303]: detent 4 spaces

File: slycot/math.py [hunk 10]
```diff
@@ -487,32 +485,25 @@ def mb03wd(job, compz, n, ilo, ihi, iloz, ihiz, H, Q, ldwork=None):
...
             H_j, j = 2, ..., p.
-
-    Q : ndarray
+    Q : array_like
             If compz = 'V', Q[:n,:n,:p] must contain the current matrix Q of
             transformations accumulated by SLICOT Library routine
             MB03VY.
             If compz = 'I', Q is ignored
-
     ldwork : int, optinal
```
[the remark this answers is not in the document]
Comment (bnavigator) [reply slycot/math.py:501]:
```suggestion
    ldwork : int, optional
```

File: slycot/math.py (hunk 7 above)
Comment (bnavigator) [reply slycot/math.py:21]:
```suggestion
from .exceptions import raise_if_slycot_error
```

File: slycot/math.py (hunk 8 above)
Comment (bnavigator) [reply slycot/math.py:537]:
```suggestion
            the elements Wr[{info}:{ihi}] contains those
```

**Comment by repagh** [comment]
Looks perfect now. Any objections to a merge?

**Comment by bnavigator** [comment]
As soon as you said it's perfect, I broke it ;) Fixed now.

With all the numpydoc changes, this is hard to review. Let's give @roryyorke a little bit of time to see if he has any objections.

File: slycot/math.py [hunk 11]
```diff
@@ -681,9 +680,9 @@ def mb05nd(a, delta, tol=1e-7):
 
     Returns
     -------
-    F : ndarray
+    F : (n n) ndarray
```
Comment (repagh) [inline slycot/math.py:683]: ,

Comment (bnavigator) [reply slycot/math.py:683]:
```suggestion
    F : (n, n) ndarray
```

**Comment by bnavigator** [comment]
So, time to merge and focus on the release before we get new ideas? ;)

**Pull request merged**